File: CSharpServer/SkyARFighter.Common/Defines.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SkyARFighter.Common
{
    public enum RemotingMethodId : int
    {
        [RemotingDirection(RemotingDirection.C2S)]
        Login = 0x01,
        [RemotingDirection(RemotingDirection.S2C)]
        Welcome = 0x02,
        [RemotingDirection(RemotingDirection.Both)]
        SetupWorld = 0x03,
        [RemotingDirection(RemotingDirection.Both)]
        SendMessage = 0x04,
        [RemotingDirection(RemotingDirection.Both)]
        CreateSceneModel = 0x10,
        [RemotingDirection(RemotingDirection.Both)]
        MoveSceneModel = 0x11,
        [RemotingDirection(RemotingDirection.Both)]
        ScaleSceneModel = 0x12,
        [RemotingDirection(RemotingDirection.Both)]
        RotateSceneModel = 0x13,
        [RemotingDirection(RemotingDirection.Both)]
        DeleteSceneModel = 0x14,
        [RemotingDirection(RemotingDirection.S2C)]
        AddPlayer = 0x20,
        [RemotingDirection(RemotingDirection.S2C)]
        RemovePlayer = 0x21,
        [RemotingDirection(RemotingDirection.Both)]
        SyncPlayerState = 0x22,
    }

    public enum LoginWay
    {
        DeviceId,
        Account,
        AccessToken
    }

    public enum SceneModelType
    {
        标注_圆点 = 0,
		标注_圆圈,
        标注_箭头,
        标注_连线,
        平面_贴图 = 10,
        平面_视频,
        平面_挡板,
        音效 = 20,
        特效,
        按钮触发器,
        角色 = 100
    }

    public enum MessageType
    {
        System_Failure,
        System_Notification,
        Player_Chat,
    }

    public enum RemotingDirection
    {
        C2S,
        S2C,
        Both
    }
    public class RemotingDirectionAttribute : Attribute
    {
        public RemotingDirectionAttribute(RemotingDirection dir)
        {
            Direction = dir;
        }

        public RemotingDirection Direction { get; private set; }
    }

}

File: ARFighter/ARFighter/Defines.swift
//
//  Defines.swift
//  ARFighter
//
//  Created by Jones on 2018/5/5.
//  Copyright © 2018年 tjwd. All rights reserved.
//

import SceneKit

enum RemotingMethodId: Int32 {
    case Login = 0x01
    case Welcome
    case SetupWorld
    case SendMessage
    case CreateSceneModel = 0x10
    case MoveSceneModel
    case ScaleSceneModel
    case RotateSceneModel
    case DeleteSceneModel
    case AddPlayer = 0x20
    case RemovePlayer
    case SyncPlayerState
}

enum PlayerState {
    case Initial
    case SearchOrigin
    case SceneLoading
    case ScenePlaying
}

enum LoginWay: Int32 {
    case DeviceId
    case Account
    case AccessToken
}

enum OperationMode: Int32 {
    case None
    case Create
    case Delete
    case Move
    case Rotate
    case Scale
}

class GameObject {
    
    var id: Int64 = generateAutoId()
    
    static func generateAutoId() -> Int64 {
        let ct = Calendar.current
        let coms = ct.dateComponents([.year, .month, .day, .hour, .minute, .second], from: Date())
        let tail = arc4random_uniform(100000)
        let idstr = String(format: "%04d%02d%02d%02d%02d%02d%05d", coms.year!, coms.month!, coms.day!, coms.hour!, coms.minute!, coms.second!, tail)
        return Int64(idstr)!
    }
}

protocol MessageViewDelegate {
    func appendLog(msg: String)
    func alert(msg: String)
    func updatePlayerState(name: String?, state: PlayerState)
}

protocol RemotingMethodDelegate {
    func invokeMethod(methodId: RemotingMethodId, args: [Any])
}


